src/replication.c

```diff
@@ -2764,6 +2713,215 @@ cleanup_deadlist(void *arg)
 	return (NULL);
 }
 
+static int
+enqueue_mgmt_command_to_replica(replica_t *replica, uint64_t io_seq, uint8_t *data,
+    uint64_t data_len, zvol_op_code_t opcode, rcommon_mgmt_cmd_t *rcomm_mgmt)
+{
+	zvol_io_hdr_t *rmgmtio = NULL;
+	mgmt_cmd_t *mgmt_cmd;
+	uint8_t *cmd_data;
+	uint64_t num = 1;
+	int ret = 0;
+
+	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+	mgmt_cmd->rcomm_mgmt = rcomm_mgmt;
+
+	BUILD_REPLICA_MGMT_HDR(rmgmtio, opcode, data_len);
+
+	cmd_data = (uint8_t *)malloc(data_len);
+	memcpy(cmd_data, data, data_len);
+
+	rmgmtio->io_seq = io_seq;
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = cmd_data;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	//TODO: Add it as the second IO, rather than tailing
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if (rcomm_mgmt != NULL)
+		rcomm_mgmt->cmds_sent++;
+
+	if (write(replica->mgmt_eventfd1, &num, sizeof (num)) != sizeof (num)) {
+		REPLICA_ERRLOG("Failed to inform to mgmt_eventfd for "
+		    "replica(%lu)\n", replica->zvol_guid);
+		ret = -1;
+	}
+
+	return ret;
+}
+
+static int
+send_io_blocking_mgmt_cmd(spec_t *spec, rcommon_mgmt_cmd_t *rcomm_mgmt_cmd,
+    zvol_op_code_t opcode, int io_wait_time, int wait_time, int *mgmt_io_seq)
+{
+	replica_t *replica;
+	struct timespec last, now, diff;
+	int ret = -1, rc = -1;
+	uint64_t io_seq = 0;
+
+	ASSERT(spec);
+	ASSERT(rcomm_mgmt_cmd);
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+	MTX_LOCK(&spec->rq_mtx);
+
+	/* Wait for any ongoing mgmt commands which could block IOs. */
+	while (spec->quiesce == 1) {
+		MTX_UNLOCK(&spec->rq_mtx);
+		sleep(1);
+		MTX_LOCK(&spec->rq_mtx);
+	}
+
+	if (is_volume_healthy(spec) == false) {
+		REPLICA_ERRLOG("volume is not healthy..\n");
+		goto out;
```

Comment:
Memory leak of rcomm_mgmt_cmd if follow goto statement ?

Reply:
Yes. At this place, `rcomm_mgmt_cmd` hasn't been sent to replica so the caller should handle freeing rcomm_mgmt_cmd in this case.
 Thanks for catching this.